### PR TITLE
Bugfix/775 var context json size 0 dims

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Given that changes to class `stan::io::var_context` in branch https://github.com/stan-dev/stan/tree/feature/2687-var-context-validate-dims, PR https://github.com/stan-dev/stan/pull/2869 have been merged into Stan, this PR adds implementation of `var_context` virtual function `validate_dims` to  class `cmdstan::json`.

#### Intended Effect:
Given Stan program with data block:

```
data { 
  int d;
  vector[d] d_v1;
  vector[d] d_v2;
  row_vector[d] d_rv1;
  row_vector[d] d_rv2;
  vector[3] d_v_ar[d];
  row_vector[3] d_rv_ar[d];
  matrix[0, 3] d_m;
} 
```

accept as valid JSON input:

```
{
    "d" : 0,
    "d_v1" : [],
    "d_v2" : [],
    "d_rv1" : [],
    "d_rv2" : [],
    "d_v_ar" : [],
    "d_rv_ar" : [],
    "d_m" : [],
    "d_m_ar" : []
}
```

#### How to Verify:
unit tests

#### Side Effects:
none

#### Documentation:
added subsection to CmdStan manual illustrating JSON notation for arrays where any dimension is size 0.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by mitzimorris** [comment]
hi @rok-cesnovar, this PR is dependent on the Stan PR.  once this fix is merged into CmdStan,  CmdStanR can use JSON format, so this will remove the dependency on RStan.

**Comment by rok-cesnovar** [comment]
Thank you! Removing the rstan dependency will be very nice indeed!

I will restart the Jenkins tests in such a way that this PR will use the Stan PR code. And also review both PRs asap.

File: src/cmdstan/io/json/json_data.hpp [hunk 1]
```diff
@@ -203,6 +203,68 @@ namespace cmdstan {
         return (vars_i_.erase(name) > 0)
           || (vars_r_.erase(name) > 0);
       }
+
+  void validate_dims(const std::string& stage, const std::string& name,
+                     const std::string& base_type,
+                     const std::vector<size_t>& dims_declared) const {
+    std::cout << "validate dims" << std::endl;
```
Comment (rok-cesnovar) [inline src/cmdstan/io/json/json_data.hpp]: This is probably leftover debug prints?

File: src/cmdstan/io/json/json_data.hpp [hunk 2]
```diff
@@ -203,6 +203,68 @@ namespace cmdstan {
...
+    }
+
+    std::vector<size_t> dims = dims_r(name);
+
+    // JSON '[ ]' is ambiguous - any multi-dim variable with len 0 dim
+    size_t num_elements = 1;
+    for (size_t i = 0; i < dims.size(); ++i) {
+      num_elements *= dims[i];
+    }
+    std::cout << "num_elements: " << num_elements << std::endl;
```
Comment (rok-cesnovar) [inline src/cmdstan/io/json/json_data.hpp]: As well as this?

**Comment by rok-cesnovar** [comment]
: It also seems that the tests are passing if we use the Stan PR branch: https://jenkins.mc-stan.org/blue/organizations/jenkins/CmdStan/detail/PR-787/2/pipeline/45

**Comment by mitzimorris** [comment]
thanks @rok-cesnovar !  removed the stray debug comments.

**Pull request merged**